### PR TITLE
[mypy] Fix type annotations for maths directory

### PR DESCRIPTION
**Pull request by Rohanrbharadwaj**
### Describe your change:

* [ ] Add an algorithm?
* [x] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [ ] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [x] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

**Comment by Rohanrbharadwaj** [comment]
15 to go!

**Comment by Rohanrbharadwaj** [comment]
15 more to go!

**Comment by Rohanrbharadwaj** [comment]
@cclauss What is a merge conflict? Why am I seeing it and what should I do?

I know that mypy.ini was updated after I opened this PR in #5789. Has that got anything to do with this?

File: maths/proth_number.py [hunk 1]
```diff
@@ -44,18 +38,16 @@ def proth(number: int) -> int:
...
     else:
-        block_index = number // 3
         """
         +1 for binary starting at 0 i.e. 2^0, 2^1, etc.
         +1 to start the sequence at the 3rd Proth number
         Hence, we have a +2 in the below statement
         """
-        block_index = math.log(block_index, 2) + 2
-        block_index = int(block_index)
+        block_index: int = int(math.log(number // 3, 2)) + 2
```
Comment (cclauss) [inline maths/proth_number.py:46]: Not needed.
```suggestion
        block_index = int(math.log(number // 3, 2)) + 2
```

File: maths/proth_number.py [hunk 2]
```diff
@@ -44,18 +38,16 @@ def proth(number: int) -> int:
...
 
-        proth_list = [3, 5]
-        proth_index = 2
-        increment = 3
+        proth_list: list[int] = [3, 5]
+        proth_index: int = 2
+        increment: int = 3
```
Comment (cclauss) [inline maths/proth_number.py:50]: Not needed.

File: maths/proth_number.py [hunk 3]
```diff
@@ -66,12 +58,6 @@ def proth(number: int) -> int:
...
-    for number in range(11):
-        value = 0
-        try:
-            value = proth(number)
-        except ValueError:
-            print(f"ValueError: there is no {number}th Proth number")
-            continue
+    import doctest
 
-        print(f"The {number}th Proth number: {value}")
```
Comment (cclauss) [inline maths/proth_number.py:77]: Why remove user interaction?!?

File: maths/series/geometric_series.py [hunk 4]
```diff
@@ -11,53 +10,54 @@
...
     :return: The Geometric Series starting from first term a and multiple of common
         ration with first term with increase in power till last term (nth term)
+
     Examples:
-    >>> geometric_series(4, 2, 2)
-    [2, '4.0', '8.0', '16.0']
     >>> geometric_series(4.0, 2.0, 2.0)
-    [2.0, '4.0', '8.0', '16.0']
+    [2.0, 4.0, 8.0, 16.0]
+    >>> geometric_series(4.0, 2.0, 2.0)
```
Comment (cclauss) [inline maths/series/geometric_series.py:28]: It is still useful to test what happens is the caller ignores the type hints.  They _often_ will.
```suggestion
    >>> geometric_series(4, 2, 2)
```

File: maths/series/geometric_series.py [hunk 5]
```diff
@@ -11,53 +10,54 @@
...
+    [2.0, 4.0, 8.0, 16.0]
     >>> geometric_series(4.1, 2.1, 2.1)
-    [2.1, '4.41', '9.261000000000001', '19.448100000000004']
-    >>> geometric_series(4, 2, -2)
-    [2, '-4.0', '8.0', '-16.0']
-    >>> geometric_series(4, -2, 2)
-    [-2, '-4.0', '-8.0', '-16.0']
-    >>> geometric_series(-4, 2, 2)
```
Comment (cclauss) [inline maths/series/geometric_series.py:32]: Same.  Please revert these changes.

Comment (Rohanrbharadwaj) [reply maths/series/geometric_series.py:32]: Ok, so make the test floats ints again?

File: maths/series/geometric_series.py [hunk 6]
```diff
@@ -11,53 +10,54 @@
...
+    >>> geometric_series(1.0, 1.0, 1.0)
+    [1.0]
     >>> geometric_series(0, 0, 0)
     []
     """
-    if "" in (nth_term, start_term_a, common_ratio_r):
-        return ""
-    series = []
-    power = 1
+    if 0 in (nth_term, start_term_a, common_ratio_r):
```
Comment (cclauss) [inline maths/series/geometric_series.py:45]: Let's test for all falsey values that the caller may provide.
```suggestion
    if not all((nth_term, start_term_a, common_ratio_r)):
```

File: maths/series/geometric_series.py [hunk 7]
```diff
@@ -11,53 +10,54 @@
...
+        return []
+    series: list[float] = []
+    power: int = 1
```
Comment (cclauss) [inline maths/series/geometric_series.py:48]: Not needed.

File: maths/proth_number.py (hunk 3 above)
Comment (Rohanrbharadwaj) [reply maths/proth_number.py:77]: CONTRIBUTING.md said no print statements, but I'll revert this too

File: maths/series/p_series.py [hunk 8]
```diff
@@ -1,48 +1,38 @@
...
+    for temp in range(nth_term):
+        series.append(1 / (pow(temp + 1, power)) if series else 1.0)
     return series
 
 
 if __name__ == "__main__":
-    nth_term = input("Enter the last number (nth term) of the P-Series")
-    power = input("Enter the power for  P-Series")
-    print("Formula of P-Series => 1+1/2^p+1/3^p ..... 1/n^p")
-    print(p_series(nth_term, power))
```
Comment (cclauss) [inline maths/series/p_series.py:48]: Why remove user interaction?!?

File: maths/series/p_series.py [hunk 9]
```diff
@@ -1,48 +1,38 @@
...
     :return: The P-Series starting from 1 to last (nth) term
-
     Examples:
     >>> p_series(5, 2)
-    [1, '1/4', '1/9', '1/16', '1/25']
+    [1.0, 0.25, 0.1111111111111111, 0.0625, 0.04]
     >>> p_series(-5, 2)
     []
     >>> p_series(5, -2)
-    [1, '1/0.25', '1/0.1111111111111111', '1/0.0625', '1/0.04']
```
Comment (cclauss) [inline maths/series/p_series.py:26]: This looks like a cooler test.

Comment (Rohanrbharadwaj) [reply maths/series/p_series.py:26]: Make all strings here instead of floats, then?

File: maths/average_mode.py [hunk 10]
```diff
@@ -1,34 +1,28 @@
-def mode(input_list: list) -> list:  # Defining function "mode."
+from collections import Counter
```
Comment (cclauss) [inline maths/average_mode.py:1]: I think the point was to show how to do it, not to hide how to do it inside a Counter.

Comment (Rohanrbharadwaj) [reply maths/average_mode.py:1]: Alright, I'll keep the old method

File: maths/gamma_recursive.py [hunk 11]
```diff
@@ -71,8 +71,3 @@ def test_gamma() -> None:
     from doctest import testmod
 
     testmod()
-    num = 1
-    while num:
-        num = float(input("Gamma of: "))
-        print(f"gamma({num}) = {gamma(num)}")
-        print("\nEnter 0 to exit...")
```
Comment (cclauss) [inline maths/gamma_recursive.py:78]: Why remove user interaction?!?

File: maths/proth_number.py (hunk 3 above)
Comment (cclauss) [reply maths/proth_number.py:77]: Read closer.
1. This is NOT a function or method.
2. This is not part of the algorithm.

**Comment by Rohanrbharadwaj** [comment]
Are we supposed to have user interaction or no print statements (according to CONTRIBUTING.md)??

File: maths/series/geometric_series.py (hunk 5 above)
Comment (cclauss) [reply maths/series/geometric_series.py:32]: The idea is to MIX so 4 as an int and 4.41 as a float and make sure the algorithm does the right thing.

**Comment by Rohanrbharadwaj** [comment]
The tests still don't run because of a "merging conflict". What's that?

File: maths/proth_number.py (hunk 3 above)
Comment (Rohanrbharadwaj) [reply maths/proth_number.py:77]: Right, I got you

File: maths/average_mode.py [hunk 12]
```diff
@@ -19,10 +18,15 @@ def mode(input_list: list) -> list[Any]:  # Defining function "mode."
     >>> mode(["x", "x" , "y", "y", "z"])
     ['x', 'y']
     """
-    counts: list[int] = list((Counter(input_list)).values())
-    return list(
-        sorted({val for val in input_list if input_list.count(val) == max(counts)})
-    )
+    result = list()  # Empty list to store the counts of elements in input_list
+    for x in input_list:
+        result.append(input_list.count(x))
```
Comment (cclauss) [inline maths/average_mode.py]: Let's make this a list comprehension.

File: maths/average_mode.py [hunk 13]
```diff
@@ -19,10 +18,15 @@ def mode(input_list: list) -> list[Any]:  # Defining function "mode."
...
-    )
+    result = list()  # Empty list to store the counts of elements in input_list
+    for x in input_list:
+        result.append(input_list.count(x))
+    if not result:
+        return []
+    y = max(result)  # Gets the maximum value in the result list.
+    # Gets values of modes
+    result = list({input_list[i] for i, value in enumerate(result) if value == y})
+    return sorted(result)
```
Comment (cclauss) [inline maths/average_mode.py]:
```suggestion
    return sorted({input_list[i] for i, value in enumerate(result) if value == y})
```

File: maths/proth_number.py [hunk 14]
```diff
@@ -58,6 +58,12 @@ def proth(number: int) -> int:
 
 
 if __name__ == "__main__":
-    import doctest
+    for number in range(11):
```
Comment (cclauss) [inline maths/proth_number.py]: Let's test the code first and then let the user interact with it.

Best of both worlds!
```suggestion
    import doctest
    
    doctest.testmod()
    for number in range(11):
```

File: maths/average_mode.py [hunk 15]
```diff
@@ -1,34 +1,29 @@
-def mode(input_list: list) -> list:  # Defining function "mode."
+from typing import Any
+
+
+def mode(input_list: list) -> list[Any]:  # Defining function "mode."
```
Comment (cclauss) [inline maths/average_mode.py:4]: The comment says exactly what the code says so it can be removed.
```suggestion
def mode(input_list: list) -> list[Any]:
```

File: maths/series/p_series.py (hunk 9 above)
Comment (cclauss) [reply maths/series/p_series.py:26]: The current implementation is tolerant of both.  We should not lose that flexibility.  Our type hint says that we _expect_ float but our implementation is flexible to do the right thing if someone hands us something that is not a float but is still reasonable.

Comment (cclauss) [reply maths/series/p_series.py:26]: https://pydantic-docs.helpmanual.io does a nice job of this.

**Comment by Rohanrbharadwaj** [comment]
I'll fix that, thanks! What did you just do, though?

File: maths/series/geometric_series.py [hunk 16]
```diff
@@ -10,11 +10,14 @@
 """
 
 
+from typing import Union
```
Comment (cclauss) [inline maths/series/geometric_series.py:13]:
```suggestion
from __future__ import annotations
```
And then use the new `|` syntax below.

File: maths/series/geometric_series.py [hunk 17]
```diff
@@ -45,7 +48,7 @@ def geometric_series(
     """
     if not all((nth_term, start_term_a, common_ratio_r)):
         return []
-    series = []
+    series: list[Union[float, int]] = []
```
Comment (cclauss) [inline maths/series/geometric_series.py:51]:
```suggestion
    series: list[float | int] = []
```

File: maths/series/geometric_series.py (hunk 16 above)
Comment (Rohanrbharadwaj) [reply maths/series/geometric_series.py:13]: Alright

File: maths/series/geometric_series.py [hunk 18]
```diff
@@ -62,11 +65,3 @@ def geometric_series(
...
 
     doctest.testmod()
-
-    nth_term = float(input("Enter the last number (n term) of the Geometric Series"))
-    start_term_a = float(input("Enter the starting term (a) of the Geometric Series"))
-    common_ratio_r = float(
-        input("Enter the common ratio between two terms (r) of the Geometric Series")
-    )
-    print("Formula of Geometric Series => a + ar + ar^2 ... +ar^n")
-    print(geometric_series(nth_term, start_term_a, common_ratio_r))
```
Comment (cclauss) [inline maths/series/geometric_series.py]: Please restore.

File: maths/series/p_series.py [hunk 19]
```diff
@@ -10,29 +10,45 @@
 """
 
 
-def p_series(nth_term: int, power: int) -> list[float]:
+from typing import Union
```
Comment (cclauss) [inline maths/series/p_series.py:13]:
```suggestion
from __future__ import annotations
```

File: maths/series/p_series.py [hunk 20]
```diff
@@ -10,29 +10,45 @@
...
+
+
+def p_series(
+    nth_term: Union[int, float, str], power: Union[int, float, str]
```
Comment (cclauss) [inline maths/series/p_series.py:17]:
```suggestion
    nth_term: [int | float | str], power: [int | float | str]
```

File: maths/series/p_series.py [hunk 21]
```diff
@@ -10,29 +10,45 @@
...
+) -> Union[list[str], str]:
```
Comment (cclauss) [inline maths/series/p_series.py]: Let's not confuse the caller.  Let's pick one and only deliver them a single data type.  I think a list of strings even if that list only has one string is the least surprising.

Comment (Rohanrbharadwaj) [reply maths/series/p_series.py]: So put that string in [ ] ?

**Comment by Rohanrbharadwaj** [comment]
`data_structures/stacks/next_greater_element.py` can be removed from the excluded ones in mypy.ini, since it was fixed in #5763.

**Review by cclauss — APPROVED** [review]
AWESOME!!!
